Final integration

Kshitij1K opened the pull request:
Final changes to be merged into master.
Status, as of now:-
- Tracking and landing with/without trajectory generation is working, for reasonable velocities of husky, and also for reasonable wind velocities. This utilizes ground truth for both the quad's and husky's odometry.
- The code for using detection and pose estimation to estimate odometry also exists, however landing using that is currently not functional due to inaccuracies. This will be fixed later on, using some filter to accurately estimate pose.
- A basic README having instructions and a wiki having a very basic description has also been made. This will be made more dense in the upcoming days.
- Apart from this, there are some more tasks that will be done, later on. A list for them can be found in the README of future_tasks branch.

Kshitij1K committed:
Changed landing

Kshitij1K committed:
Changed trajectory generation

Kshitij1K committed:
Added landing for trajectory_generation_tracker

Kshitij1K committed:
Changed launch file

Kshitij1K committed:
removed unnecessary print statements

Kshitij1K committed:
Added rqt_gui support

Kshitij1K committed:
Merge branch 'master' into Integration

Kshitij1K committed:
Merge remote-tracking branch 'origin/master' into Integration

Kshitij1K committed:
Tasks

Kshitij1K committed:
Some more tasks

Kshitij1K committed:
Organized nodes/ launch files in different packages

Kshitij1K committed:
Fixed wrong placement

Kshitij1K committed:
Some more file movements

Kshitij1K committed:
CMakeLists.txt and package.xml for each package

Kshitij1K committed:
Changed include file path and removed unnecessary code

Sanjayso27 committed:
parameter cleanup and used single camera.yaml for both detection and pose estimation

Sanjayso27 committed:
some more cleanup

Kshitij1K committed:
Merge pull request #16 from AerialRobotics-IITK/parameter-cleanup

Parameter cleanup

Kshitij1K committed:
Moved parameter file

Kshitij1K committed:
Added random husky trajectory, Updated task list

Kshitij1K committed:
Changed launch files

Kshitij1K committed:
Renamed and added launch files

Kshitij1K committed:
Updated tasks

Kshitij1K committed:
Fixed flaw in parameter file

Kshitij1K committed:
made launch files better

Kshitij1K committed:
pose estimation now publishes nav_msgs::Odometry

Kshitij1K committed:
Fixed erroneous getParam command

Kshitij1K committed:
Extra tasks

Kshitij1K committed:
extra Params

Kshitij1K committed:
Added support for using detection

Kshitij1K committed:
Minor fixes

Kshitij1K committed:
Added support for using detection

Sanjayso27 committed:
added teleop twistr keyboard

Kshitij1K committed:
teleop Twist keyboard support

Kshitij1K committed:
Added support for using detection

Kshitij1K committed:
Added ifplatformdetected check

Kshitij1K committed:
Renamed package

Kshitij1K committed:
Changed folder name

Kshitij1K committed:
Created package and written basic backbone

Kshitij1K committed:
Added calculations

Kshitij1K committed:
Fixed minor error

Kshitij1K committed:
Task added

Kshitij1K committed:
Files renames

Kshitij1K committed:
Used pose_estimation_lib

Kshitij1K committed:
Camera Params changed

varunk122 committed:
change name husky to platform, add wind , add params

varunk122 committed:
fixed model

Kshitij1K committed:
Fixed erroneous params

Kshitij1K committed:
Merge branch 'improved_pose_estimation' into Integration

varunk122 committed:
add wind, changed pkg name to pose_est_ros,added pose_est

varunk122 committed:
made some changes in pose_estimation_ros

varunk122 committed:
made a change

Kshitij1K committed:
Final changes

Kshitij1K committed:
Merge branch 'Integration' into final

Kshitij1K committed:
Merge pull request #18 from AerialRobotics-IITK/final

Final

Kshitij1K committed:
Minor fix

Kshitij1K committed:
Fixed more errors

Kshitij1K committed:
Added custom husky files

Kshitij1K committed:
Husky with custom top plate

Kshitij1K committed:
Configured launch files to use custom husky

Kshitij1K committed:
Final debugging

Kshitij1K committed:
Added platform_detected condition

Kshitij1K committed:
Checked files for conventions

Kshitij1K committed:
Added rosmon and finished rqt_gui

Kshitij1K committed:
Added a basic instructions

Sanjayso27 reviewed (approved):
kshitij I am reminding u for adding the installation instructions of teleop twist keyboard in readme
and I have only one doubt that the hsv parameters for thresholding was changed because u added the red patform right?

Kshitij1K committed:
Added instructions for using teleop twist keyboard

Kshitij1K committed:
Added install instructions for husky_gazebo

Kshitij1K commented:
Added instructions for using teleop twist keyboard, and also for installing husky_simulator.
Yes, I changed the hsv thresholding parameters because now we are using a red top plate for detection, to differentiate it from the black wheels and shadow, that is causing it to inaccurately calculate the bounding box and give erroneous centre coordinates of platform.

ashwin2802 commented on auto_landing/cfg/camera_params.yaml:
```diff
@@ -1,9 +1,17 @@
-camera_to_quad_matrix: [-1, 0, 0, 0, -1, 0, 0, 0, -1]
-camera_translation: [0.20, 0.0, -0.02]
+##Camera Parameters
+distortion_matrix:                                                                      #double matrix(1x5): used for undistorting distorted images
```
that seems to be a lotta space though

ashwin2802 commented on auto_landing/cfg/camera_params.yaml:
```diff
@@ -1,9 +1,17 @@
+  rows: 3
+  cols: 3
+  data: [0, -1, 0, -1, 0, 0, 0, 0, -1]
+camera_translation:                                                                     #double matrix(1,3): translation vector from quad centre to the camera.
```
plis mention units of parameters

ashwin2802 commented on detection/include/detection/detect_platform.hpp:
```diff
@@ -3,21 +3,21 @@
 
-namespace ariitk::auto_landing {
+namespace ariitk::detection {
 class PlatformDetect {
```
space here

Kshitij1K replied:
Where? Do you mean a newline after the namespace?

ashwin2802 replied:
sorry, yes i meant line

Kshitij1K replied:
Where?

ashwin2802 commented on detection/include/detection/detect_platform.hpp:
```diff
@@ -29,13 +29,14 @@ class PlatformDetect {
 		int contour_perimeter_threshold_, contour_perimeter_scale_;
 		int kernel_size_;
 		geometry_msgs::Point center_;
+		std_msgs::Bool is_platform_detected;
```
end with underscore

ashwin2802 commented on pose_estimation/include/pose_estimation/pose.hpp:
```diff
@@ -0,0 +1,41 @@
+
+class PoseEstimation {
+	public:
+                PoseEstimation(){};
```
is this the right amount of indenting?

ashwin2802 commented on pose_estimation/include/pose_estimation/pose.hpp:
```diff
@@ -0,0 +1,41 @@
+                Eigen::Vector3d pixel_coordinates_;
+
+                bool is_data_changed_;
+                enum use_case { scaleup_known, z_known } estimation_type;
```
prefer using an `enum class` instead, as this will protect the enum constants, otherwise they can be directly accessed within the namespace

ashwin2802 commented on pose_estimation/src/pose.cpp:
```diff
@@ -0,0 +1,82 @@
+void PoseEstimation::scaleupKnownCalculation() {
+	object_position_ =
+	    (vehicle_to_ground_rotation_ *
+	        ((camera_to_vehicle_rotation_ * ((scale_up_matrix_ * inverse_camera_matrix_ * pixel_coordinates_) + camera_centre_displacement_)) +
```
can we make this look a little nicer by using intermediate storage variables?

ashwin2802 commented on pose_estimation_ros/src/pose_estimation_ros.cpp:
```diff
@@ -0,0 +1,86 @@
+	scaleUpMatrix = Eigen::Matrix3d::Zero();
+
+	platform_odom_[0].pose.pose.position.x = 0;
+	platform_odom_[0].pose.pose.position.y = 0;
```
better impl: `platform_odom_[0].pose.pose.position = {0.0, 0.0, 0.0}`

Kshitij1K replied:
This gives build errors.

ashwin2802 replied:
yeah, i wasn't sure if it would work. no prob

ashwin2802 commented on pose_estimation_ros/src/pose_estimation_ros.cpp:
```diff
@@ -0,0 +1,86 @@
+
+void PoseEstimationROS::pixelCoordinatesCallBack(const geometry_msgs::Point& msg) { pixel_coordinates_ = msg; }
+
+void PoseEstimationROS::platformStatusCallback(const std_msgs::Bool& msg) { is_platform_detected_ = msg; }
```
can put oneliners in header

Kshitij1K replied:
I prefer putting all functions in the cpp file, in case I have to do some processing in them later on.

ashwin2802 replied:
cool, just see that then there are no definitions in headers to maintain consistency

ashwin2802 commented on tracking/include/tracking/tracking.hpp:
```diff
@@ -0,0 +1,42 @@
+		void init(ros::NodeHandle& nh, ros::NodeHandle& nh_private, char** argv);
+		void run();
+
+		double publish_rate_;
```
public variable wont have ending underscore

ashwin2802 commented on tracking/include/tracking/trajectory_generation_tracker.hpp:
```diff
@@ -0,0 +1,57 @@
+		TrajectoryGenerationTracking(char** argv);
+		void run();
+
+		int publish_rate_;
```
public variable wont have ending underscore

ashwin2802 commented on tracking/include/tracking/trajectory_generation_tracker.hpp:
```diff
@@ -0,0 +1,57 @@
+
+		visualization_msgs::MarkerArray markers_;
+
+		nav_msgs::Odometry mav_odom_, platform_odom_;
```
one variable per line has been followed in other files

ashwin2802 commented on tracking/src/tracking.cpp:
```diff
@@ -0,0 +1,119 @@
+
+void Tracking::run() {
+	if (!using_detection_) {
+		if ((fabs(platform_cmd_vel_[0].linear.x) > 0.0001 || fabs(platform_cmd_vel_[0].linear.y) > 0.0001)) { updateSetPoint(); }
```
think an inline `isPlatformMoving()` would look better, can also parametrize `0.0001`

Kshitij1K replied:
Yeah, also the values approximation formula used in the setpoint calculation is pretty dirty, can be moved to a separate function.

Kshitij1K replied:
Also, the 0.00001 isn't any functional value having any use, it is just to compare double numbers to 0.

ashwin2802 commented on tracking/src/tracking.cpp:
```diff
@@ -0,0 +1,119 @@
+	platform_odom_.pose.pose.position.z = msg.pose[index].position.z;
+
+	platform_cmd_vel_[0].linear.x = platform_cmd_vel_[1].linear.x;
+	platform_cmd_vel_[0].linear.x = platform_cmd_vel_[1].linear.x;
```
same line thrice?

ashwin2802 commented on tracking/src/tracking.cpp:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	platform_odom_.pose.pose.position.x = msg.pose[index].position.x;
+	platform_odom_.pose.pose.position.y = msg.pose[index].position.y;
```
`platform_odom_.pose.pose.position = msg.pose[index].position` should work

ashwin2802 replied:
same for the next two blocks

ashwin2802 commented on tracking/src/tracking.cpp:
```diff
@@ -0,0 +1,119 @@
+	std::string name = msg.name[index];
+	while (name != "/") {
+		index++;
+		name = msg.name[index];
```
oneliner `while (name != "/") { name = msg.name[++index]; }`

ashwin2802 commented on tracking/src/tracking.cpp:
```diff
@@ -0,0 +1,119 @@
+	double distance_y = platform_odom_.pose.pose.position.y - quad_odom_.pose.pose.position.y;
+	double norm = sqrt(pow(distance_x, 2) + pow(distance_y, 2));
+
+	if ((fabs(distance_x) > max_x_) || (fabs(distance_y) > max_y_)) {
```
can replace this with an inline `isPlatformNear()`

ashwin2802 commented on tracking/src/trajectory_generation_tracker_node.cpp:
```diff
@@ -0,0 +1,25 @@
+
+	int count = 0;
+	while (count < 10) {
+		ros::spinOnce();
```
is this required? can convert into `tracking.waitForStartup()`

Kshitij1K replied:
Yes this is required, to ensure that odometry message has been recieved in tracking.
Will make it a function

ashwin2802 reviewed (changes requested):
Launch files indents are messed up again. Should we have to manually fix this everytime?
Also, add a LICENSE. Some of the Clearpath files use BSD, the RotorS files use Apache 2.0, so please see which is compatible (possibly MIT or GPL would do). Remember to add them to the `package.xml` files too
Please specify units of all params
Do you plan to add inline comments in the libraries?

Kshitij1K committed:
Review changes

Kshitij1K committed:
Fixed build errors

amartyadash commented on tracking/launch/tracker_sim.launch:
```diff
@@ -0,0 +1,35 @@
+    <group unless="$(arg use_trajectory_generation)" >
+    	<node pkg="tracking" type="tracking_tracking_node" name="tracking_tracking_node" output="screen" args="-use_odometry_for_platform $(arg use_odometry_for_platform)">
+            <rosparam command="load" file="$(arg tracking_parameters)" />
+            <remap from="command_pose" 	         to="$(arg mav_name)/command/track/pose" />
```
align the remaps properly here

Kshitij1K committed:
Added license to package.xml

Kshitij1K committed:
Improper indentation fixed, newlines added

Kshitij1K commented:
Launch files have been messed because some of us have been using tabs for indent, whereas others are using spaces for indents. Since GitHub displays tabs as 8 spaces long, this cause it to look like that. So for next time, everyone should use spaces for indent (so that the whitespace length is not editor dependent). Also a GNU GPL v3.0 license has been added, please also check out the pull request for adding the LICENSE.md file(package.xml has been changed in this PR). Units have been specified (wherever meaningful). Also, I do plan on adding inline comments, will do it along with improving the documentation.
